grove/amplification/amplification.py

```diff
@@ -40,27 +52,36 @@ def amplification_circuit(algorithm, oracle, qubits, num_iter):
         subspace.
     :param Sequence qubits: the qubits to operate on
     :param int num_iter: number of iterations of amplifications to run
+    :param bool decompose_diffusion: If True, decompose the Grover diffusion gate into two qubit
+     gates. If False, use a defgate to define the gate.
     :return: The amplified algorithm.
     :rtype: Program
     """
-    if num_iter <= 0:
-        raise ValueError("num_iter must be greater than zero")
-
     prog = pq.Program()
 
     uniform_superimposer = pq.Program().inst([H(qubit) for qubit in qubits])
     prog += uniform_superimposer
-
+    diffusion_program = diffusion_circuit(qubits)
+    # To avoid redefining gates, we collect them before building our program.
+    defined_gates = oracle.defined_gates + algorithm.defined_gates + diffusion_program.defined_gates
     for _ in range(num_iter):
-        prog += oracle + algorithm.dagger() + diffusion_program(qubits) + algorithm
+        prog += (oracle.instructions
+                 + algorithm.dagger().instructions
+                 + diffusion_program.instructions
+                 + algorithm.instructions)
+    # We redefine the gates in the new program.
+    for gate in defined_gates:
```

Comment:
This is fixed in the latest version of PyQuil that will be released today.

Reply:
@stevenheidel It doesn't look to me like it is, here I am being careful about how I compose programs so that I don't duplicate `defgates`. I think as it is, we still just `extend` `defgate`s when we add `Programs`.

Reply:
@stevenheidel @ampolloreno can we test if this is fixed in the new pyQuil version? This shouldn't be a point of contention.